Bottom: c8cec6da8a6ce308e9033afb5587f09e5f6d10b6
Top:    204a742b60775fa02ec358ae1cfcde75cb97a8d3
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2011-04-05 20:52:14 +0100

Refresh of give-the-hids-control-over-att

---

diff --git a/src/hid/batch/batch.c b/src/hid/batch/batch.c
index 96f73bf..1fa7ac1 100644
--- a/src/hid/batch/batch.c
+++ b/src/hid/batch/batch.c
@@ -356,6 +356,7 @@ HID batch_gui = {
   batch_invalidate_lr,
   batch_invalidate_all,
   0 /* batch_notify_crosshair_change */,
+  0 /* batch_notify_mark_change */,
   batch_set_layer,
   batch_make_gc,
   batch_destroy_gc,
diff --git a/src/hid/bom/bom.c b/src/hid/bom/bom.c
index a693efb..847eed0 100644
--- a/src/hid/bom/bom.c
+++ b/src/hid/bom/bom.c
@@ -547,6 +547,7 @@ HID bom_hid = {
   0,				/* bom_invalidate_lr */
   0,				/* bom_invalidate_all */
   0,				/* bom_notify_crosshair_change */
+  0,				/* bom_notify_mark_change */
   0,				/* bom_set_layer */
   0,				/* bom_make_gc */
   0,				/* bom_destroy_gc */
diff --git a/src/hid/common/extents.c b/src/hid/common/extents.c
index 124543d..743c81d 100644
--- a/src/hid/common/extents.c
+++ b/src/hid/common/extents.c
@@ -187,6 +187,7 @@ static HID extents_hid = {
   0 /* extents_invalidate_lr */ ,
   0 /* extents_invalidate_all */ ,
   0 /* extents_notify_crosshair_changed */ ,
+  0 /* extents_notify_mark_changed */ ,
   extents_set_layer,
   extents_make_gc,
   extents_destroy_gc,
diff --git a/src/hid/common/hidnogui.c b/src/hid/common/hidnogui.c
index 0e1a966..7c06bdf 100644
--- a/src/hid/common/hidnogui.c
+++ b/src/hid/common/hidnogui.c
@@ -421,6 +421,7 @@ HID hid_nogui = {
   nogui_invalidate_lr,
   nogui_invalidate_all,
   0 /* nogui_notify_crosshair_change */ ,
+  0 /* nogui_notify_mark_change */ ,
   nogui_set_layer,
   nogui_make_gc,
   nogui_destroy_gc,
diff --git a/src/hid/gcode/gcode.c b/src/hid/gcode/gcode.c
index f3491bd..a2cae9f 100644
--- a/src/hid/gcode/gcode.c
+++ b/src/hid/gcode/gcode.c
@@ -1044,6 +1044,7 @@ HID gcode_hid = {
   0 /* gcode_invalidate_lr */ ,
   0 /* gcode_invalidate_all */ ,
   0 /* gcode_notify_crosshair_change */ ,
+  0 /* gcode_notify_mark_change */ ,
   gcode_set_layer,
   gcode_make_gc,
   gcode_destroy_gc,
diff --git a/src/hid/gerber/gerber.c b/src/hid/gerber/gerber.c
index b8587d6..94d73e3 100644
--- a/src/hid/gerber/gerber.c
+++ b/src/hid/gerber/gerber.c
@@ -288,6 +288,7 @@ static HID gerber_hid = {
   0 /* gerber_invalidate_lr */ ,
   0 /* gerber_invalidate_all */ ,
   0 /* gerber_notify_crosshair_change */ ,
+  0 /* gerber_notify_mark_change */ ,
   gerber_set_layer,
   gerber_make_gc,
   gerber_destroy_gc,
diff --git a/src/hid/gtk/gtkhid-gdk.c b/src/hid/gtk/gtkhid-gdk.c
index ba1c4ad..95106eb 100644
--- a/src/hid/gtk/gtkhid-gdk.c
+++ b/src/hid/gtk/gtkhid-gdk.c
@@ -780,7 +780,7 @@ ghid_invalidate_all ()
 void
 ghid_notify_crosshair_change (bool changes_complete)
 {
-#warning FIXME
+  /* FIXME: This is because we sometimes get called before the GUI up */
   if (gport->drawing_area == NULL)
     return;
 
@@ -791,6 +791,20 @@ ghid_notify_crosshair_change (bool changes_complete)
   ghid_draw_area_update (gport, NULL);
 }
 
+void
+ghid_notify_mark_change (bool changes_complete)
+{
+  /* FIXME: This is because we sometimes get called before the GUI up */
+  if (gport->drawing_area == NULL)
+    return;
+
+  if (!Marked.status)
+    return;
+
+  DrawMark (); /* NB: This is an XOR drawing operation */
+  ghid_draw_area_update (gport, NULL);
+}
+
 static void
 draw_right_cross (GdkGC *xor_gc, gint x, gint y)
 {
diff --git a/src/hid/gtk/gtkhid-main.c b/src/hid/gtk/gtkhid-main.c
index 57a1b37..25ea1c3 100644
--- a/src/hid/gtk/gtkhid-main.c
+++ b/src/hid/gtk/gtkhid-main.c
@@ -989,6 +989,7 @@ HID ghid_hid = {
   ghid_invalidate_lr,
   ghid_invalidate_all,
   ghid_notify_crosshair_change,
+  ghid_notify_mark_change,
   ghid_set_layer,
   ghid_make_gc,
   ghid_destroy_gc,
diff --git a/src/hid/gtk/gui.h b/src/hid/gtk/gui.h
index ec6451f..998e2a0 100644
--- a/src/hid/gtk/gui.h
+++ b/src/hid/gtk/gui.h
@@ -495,6 +495,7 @@ void ghid_fill_rect (hidGC gc, int x1, int y1, int x2, int y2);
 void ghid_invalidate_lr (int left, int right, int top, int bottom);
 void ghid_invalidate_all ();
 void ghid_notify_crosshair_change (bool changes_complete);
+void ghid_notify_mark_change (bool changes_complete);
 void ghid_show_crosshair (gboolean show);
 void ghid_init_renderer (int *, char ***, GHidPort *);
 void ghid_init_drawing_widget (GtkWidget *widget, GHidPort *);
diff --git a/src/hid/lesstif/main.c b/src/hid/lesstif/main.c
index 3dbc6bc..9324385 100644
--- a/src/hid/lesstif/main.c
+++ b/src/hid/lesstif/main.c
@@ -1352,7 +1352,6 @@ work_area_input (Widget w, XtPointer v, XEvent * e, Boolean * ctd)
 	}
         ignore_release = 0;
 
-        HideCrosshair ();
         notify_crosshair_change (false);
         pressed_button = e->xbutton.button;
         mods = ((e->xbutton.state & ShiftMask) ? M_Shift : 0)
@@ -2880,8 +2879,8 @@ lesstif_invalidate_all (void)
   lesstif_invalidate_lr (0, PCB->MaxWidth, 0, PCB->MaxHeight);
 }
 
-void
-lesstif_notify crosshair_change (bool changes_complete)
+static void
+lesstif_notify_crosshair_change (bool changes_complete)
 {
   if (changes_complete)
     CrosshairOn ();
@@ -2889,6 +2888,14 @@ lesstif_notify crosshair_change (bool changes_complete)
     CrosshairOff ();
 }
 
+static void
+lesstif_notify_mark_change (bool changes_complete)
+{
+  if (!Marked.status)
+    return
+  DrawMark ();
+}
+
 static int
 lesstif_set_layer (const char *name, int group, int empty)
 {
@@ -3823,6 +3830,7 @@ HID lesstif_gui = {
   lesstif_invalidate_lr,
   lesstif_invalidate_all,
   lesstif_notify_crosshair_change,
+  lesstif_notify_mark_change,
   lesstif_set_layer,
   lesstif_make_gc,
   lesstif_destroy_gc,
diff --git a/src/hid/lpr/lpr.c b/src/hid/lpr/lpr.c
index fdcdb53..c932047 100644
--- a/src/hid/lpr/lpr.c
+++ b/src/hid/lpr/lpr.c
@@ -122,6 +122,7 @@ HID lpr_hid = {
   0 /* lpr_invalidate_lr */ ,
   0 /* lpr_invalidate_all */ ,
   0 /* lpr_notify_crosshair_change */ ,
+  0 /* lpr_notify_mark_change */ ,
   0 /* lpr_set_layer */ ,
   0 /* lpr_make_gc */ ,
   0 /* lpr_destroy_gc */ ,
diff --git a/src/hid/nelma/nelma.c b/src/hid/nelma/nelma.c
index e20068e..6b131ad 100644
--- a/src/hid/nelma/nelma.c
+++ b/src/hid/nelma/nelma.c
@@ -1060,6 +1060,7 @@ HID             nelma_hid = {
 	0 /* nelma_invalidate_lr */ ,
 	0 /* nelma_invalidate_all */ ,
 	0 /* nelma_notify_crosshair_change */ ,
+	0 /* nelma_notify_mark_change */ ,
 	nelma_set_layer,
 	nelma_make_gc,
 	nelma_destroy_gc,
diff --git a/src/hid/png/png.c b/src/hid/png/png.c
index 017d517..8817431 100644
--- a/src/hid/png/png.c
+++ b/src/hid/png/png.c
@@ -1528,6 +1528,7 @@ HID png_hid = {
   0 /* png_invalidate_lr */ ,
   0 /* png_invalidate_all */ ,
   0 /* png_notify_crosshair_change */ ,
+  0 /* png_notify_mark_change */ ,
   png_set_layer,
   png_make_gc,
   png_destroy_gc,
diff --git a/src/hid/ps/eps.c b/src/hid/ps/eps.c
index 03cceb6..97f9d41 100644
--- a/src/hid/ps/eps.c
+++ b/src/hid/ps/eps.c
@@ -65,6 +65,7 @@ static HID eps_hid = {
   0 /* eps_invalidate_lr */ ,
   0 /* eps_invalidate_all */ ,
   0 /* eps_notify_crosshair_change */ ,
+  0 /* eps_notify_mark_change */ ,
   eps_set_layer,
   eps_make_gc,
   eps_destroy_gc,
diff --git a/src/hid/ps/ps.c b/src/hid/ps/ps.c
index 85c3674..822f5bd 100644
--- a/src/hid/ps/ps.c
+++ b/src/hid/ps/ps.c
@@ -1354,6 +1354,7 @@ HID ps_hid = {
   0 /* ps_invalidate_lr */ ,
   0 /* ps_invalidate_all */ ,
   0 /* ps_notify_crosshair_change */ ,
+  0 /* ps_notify_mark_change */ ,
   ps_set_layer,
   ps_make_gc,
   ps_destroy_gc,
